share/nixos-shell/nixos-shell.nix

```diff
@@ -1,56 +1,156 @@
+          types.path (target: {
+            inherit target;
+          })
+          (types.submodule ({ config, ... }: {
+            options = {
+              target = mkOption {
+                type = types.path;
+                description = "Target on the guest.";
+              };
+  
+              inherit cache;
+  
+              tag = mkOption {
+                type = types.str;
+                internal = true;
+              };
+            };
+  
+            config.tag = lib.mkDefault (
+              builtins.substring 0 31 ( # tags must be shorter than 32 bytes
+                "a" + # tags must not begin with a digit
+                builtins.hashString "md5" config._module.args.name
+              )
+            );
+          }))
+        );
+        default = {};
+      };
+    };
   };
-  services.mingetty.helpLine = ''
-    Log in as "root" with an empty password.
-  '';
-
-  # build-vm overrides our filesystem settings in nixos-config
-  # cache=loose -> bad idea? Well, at least it is fast!1!!
-  boot.initrd.postMountCommands = ''
-    mkdir -p $targetRoot/home
-    mount -t 9p home $targetRoot/home -o trans=virtio,version=9p2000.L,cache=loose
-    ${lib.optionalString (user != "") ''
-      mkdir -p $targetRoot/nix/var/nix/profiles/per-user/${user}/profile/
-      mount -t 9p nixprofile $targetRoot/nix/var/nix/profiles/per-user/${user}/profile/ -o trans=virtio,version=9p2000.L,cache=loose
-    ''}
-
-    export targetRoot
-    ${pkgs.bash}/bin/bash <<\EOF
-    eval "${mounts}"
-    for mount_tag in "''${!mounts[*]}"; do
-      target=$targetRoot/"''${mounts[$mount_tag]}"
-      mkdir -p "$target"
-      mount -t 9p $mount_tag "$target" -o trans=virtio,version=9p2000.L,cache=loose
-    done
-    EOF
-  '';
-  environment.loginShellInit = ''
-    # fix terminal size
-    eval "$(${pkgs.xterm}/bin/resize)"
-
-    ${lib.optionalString (pwd != "") "cd '${pwd}' 2>/dev/null"}
-    ${lib.optionalString (term != "") "export TERM='${term}'"}
-    ${lib.optionalString (path != "") "export PATH=${path}:$PATH"}
-  '';
-
-  systemd.services."serial-getty@ttyS0".enable = true;
-  networking.firewall.enable = false;
-
-  imports = lib.optional (nixos_config != "") nixos_config;
+
+  config = let
+    user = builtins.getEnv "USER";
+    shell = builtins.baseNameOf (builtins.getEnv "SHELL");
+  in lib.mkMerge [
+    # Enable the module of the user's shell for some sensible defaults.
+    (lib.mkIf (options.programs ? ${shell}.enable) {
+      programs.${shell}.enable = mkVMDefault true;
+    })
+
+    (lib.mkIf (pkgs ? ${shell}) {
+      users.extraUsers.root.shell = mkVMDefault pkgs.${shell};
+    })
+
+    (let
+      home = builtins.getEnv "HOME";
+    in lib.mkIf (home != "" && cfg.mounts.mountHome) {
+      users.extraUsers.root.home = lib.mkVMOverride home;
+    })
+
+    # Allow passwordless ssh login with the user's key if it exists.
+    (let
+      pubkey = "${builtins.getEnv "HOME"}/.ssh/id_rsa.pub";
```

Comment:
Can you mention this in the readme?

Reply:
I would at least also check for `id_ecdsa.pub` and `id_ed25519.pub`.

Reply:
```nix
(let
  keys = lib.map (key: "${builtins.getEnv "HOME"}/.ssh/${key}")
    ["id_rsa.pub" "id_ecdsa.pub" "id_ed25519.pub"];
in {
  users.users.root.openssh.authorizedKeys.keyFiles = lib.filter builtins.pathExists keys;
})
```

Reply:
Thanks for the snippet, applied.